### ISSUE TITLE
Low performance of IMap.value when using PartitionPredicate

### ISSUE THREAD

**Issue by gkrs** [hazelcast/hazelcast#13322]
When using `PartitionPredicate` in `IMap.values` Hazelcast doesn't leverage index and the performance of such query is low. When `PartitionPredicate` is not used the query also runs slow, because whole cluster is then queried.
This is practically a follow up from #12351, where it was stated, that having PartitionPredicate with index maybe helpful, though.
If this won't be solved in near future could you please advise how to create a custom predicate to run on chosen partition with using index?

Thanks

**Comment by jerrinot** [comment]
hi @gkrs,

the issue is the index is not partitioned. 

Each member has 1 index per each map. And this index contains data for all partitions stored on this member. Hence the index cannot give you results for a single partition only. This is the reason why index is not used when your query is wrapped inside `PartitionPredicate`. 

This design has an advantage: When you require results from all partitions (=the common use-case) then it's just a single hit into an index (on each member). If we make the index partitioned then Hazelcast would have to query each (partitioned) index separately and then compose the results.

**Comment by gkrs** [comment]
Hi @jerrinot,

Thank you for quick reply. How about custom predicate mentioned in previous issue? Do you have any idea how to prepare it for partition and index combination?

**Comment by jerrinot** [comment]
hi, unfortuatenaly I have no good answer for now. 

If you know your target partition then you could use `IExecutorService` and send a Callable to the member owning the partition. The callable could call `imap.localKeySet(predicate)` without the partition predicate. It would hit only a single member (=`localKeySet()`) and I because it does not use the `PartitionPredicate` then it could uses an index. 

This would give you matching keys so then you could call `imap.get(Set<K> keys)` and send results back to the original caller. I admit it's not very elegant:) Maybe @taburet has a better idea. 

Hazelcast Enterprise uses partitioned index with HD memory and I believe `PartitionPredicate`, but that's done for different reasons.

**Comment by taburet** [comment]
Just as an idea: we may try to use global indexes for partition predicates and then filter the returned result set by comparing the target partition ID with the ID returned by `IPartitionService#getPartitionId(com.hazelcast.nio.serialization.Data)` for the key of the entry. In some cases this may speed up the execution.

**Comment by gkrs** [comment]
Thanks a lot @jerrinot @taburet, I'll definitely look into both options.
I also saw you mentioning the HD cache and later found out it is a feature in enterprise, so no help for me :disappointed:

I'll let you know if any of these improved the execution speed.

**Comment by mmedenjak** [comment]
Duplicate of:
https://github.com/hazelcast/hazelcast/issues/12351
https://github.com/hazelcast/hazelcast/issues/13563

**Comment by mmedenjak** [comment]
Since we're heavily invested in the new query engine, I don't think we should spend too much time on this.

**Comment by gdela** [comment]
@mmedenjak Do you mean there's some new mechanism, this "new query engine", that we can use instead of `PartitonPredicate`  to achieve same result as the predicate, but that will be faster? If yes, can you point me where I can read about it?

**Comment by mmedenjak** [comment]
Yes, it's now in its' initial release and we'll be releasing more of it in subsequent releases - https://docs.hazelcast.org/docs/4.1-BETA-1/manual/html-single/index.html#sql. It's a SQL query engine which is lazy by design so paging predicate is not needed, you can just cancel the query after some time.

The design details are here - https://github.com/hazelcast/hazelcast/tree/master/docs/design/sql.
